File: dev-server.js
const http = require("http");
const https = require("https");
const fs = require("fs");
const path = require("path");

const PORT = Number(process.env.PORT || 8000);
const HOST = "127.0.0.1";
const REMOTE_STOCK_URL = "https://api.growagarden2wiki.net/api/v1/games/grow-a-garden-2/stock";
const ROOT = __dirname;

const MIME = {
  ".html": "text/html; charset=utf-8",
  ".js": "application/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".txt": "text/plain; charset=utf-8",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
  ".ico": "image/x-icon"
};

function send(res, status, body, contentType = "text/plain; charset=utf-8") {
  res.writeHead(status, {
    "Content-Type": contentType,
    "Cache-Control": "no-store"
  });
  res.end(body);
}

function proxyStock(res) {
  const request = https.request(
    REMOTE_STOCK_URL,
    {
      method: "GET",
      headers: {
        Accept: "application/json, text/plain, */*",
        Origin: "https://growagarden2wiki.net",
        Referer: "https://growagarden2wiki.net/stock/",
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/148.0.0.0 Safari/537.36"
      }
    },
    (upstream) => {
      let body = "";
      upstream.setEncoding("utf8");
      upstream.on("data", (chunk) => {
        body += chunk;
      });
      upstream.on("end", () => {
        const status = upstream.statusCode || 502;
        const contentType = upstream.headers["content-type"] || "application/json; charset=utf-8";
        send(res, status, body, contentType);
      });
    }
  );

  request.on("error", (error) => {
    const payload = JSON.stringify({
      error: "proxy_request_failed",
      message: error.message
    });
    send(res, 502, payload, "application/json; charset=utf-8");
  });

  request.end();
}

function serveFile(urlPath, res) {
  const safePath = urlPath === "/" ? "/index.html" : urlPath;
  const absolute = path.join(ROOT, safePath);
  const normalized = path.normalize(absolute);

  if (!normalized.startsWith(ROOT)) {
    send(res, 403, "Forbidden");
    return;
  }

  fs.stat(normalized, (statError, stats) => {
    if (statError || !stats.isFile()) {
      send(res, 404, "Not found");
      return;
    }

    const ext = path.extname(normalized).toLowerCase();
    const contentType = MIME[ext] || "application/octet-stream";

    fs.readFile(normalized, (readError, data) => {
      if (readError) {
        send(res, 500, "Failed to read file");
        return;
      }
      res.writeHead(200, {
        "Content-Type": contentType,
        "Cache-Control": "no-store"
      });
      res.end(data);
    });
  });
}

const server = http.createServer((req, res) => {
  const url = new URL(req.url || "/", `http://${req.headers.host || "localhost"}`);

  if (url.pathname === "/api/gag2-stock.json") {
    proxyStock(res);
    return;
  }

  serveFile(url.pathname, res);
});

server.listen(PORT, HOST, () => {
  console.log(`GAG2 dev server running at http://${HOST}:${PORT}`);
  console.log("Serves static tracker files and proxies /api/gag2-stock.json");
});

File: dev-server.py
from http.server import ThreadingHTTPServer, SimpleHTTPRequestHandler
from urllib.parse import urlparse
from urllib.request import Request, urlopen
from urllib.error import URLError, HTTPError
import json
import os

HOST = "127.0.0.1"
PORT = 8000
ROOT = os.path.dirname(os.path.abspath(__file__))
REMOTE_STOCK_URL = "https://api.growagarden2wiki.net/api/v1/games/grow-a-garden-2/stock"


class Handler(SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def _send_json(self, status, payload):
        body = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(body)

    def _proxy_stock(self):
        req = Request(
            REMOTE_STOCK_URL,
            headers={
                "Accept": "application/json, text/plain, */*",
                "Origin": "https://growagarden2wiki.net",
                "Referer": "https://growagarden2wiki.net/stock/",
                "User-Agent": (
                    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                    "AppleWebKit/537.36 (KHTML, like Gecko) "
                    "Chrome/148.0.0.0 Safari/537.36"
                ),
            },
        )

        try:
            with urlopen(req, timeout=20) as response:
                body = response.read()
                status = response.getcode()
                content_type = response.headers.get("Content-Type", "application/json; charset=utf-8")
                self.send_response(status)
                self.send_header("Content-Type", content_type)
                self.send_header("Content-Length", str(len(body)))
                self.send_header("Cache-Control", "no-store")
                self.end_headers()
                self.wfile.write(body)
        except HTTPError as err:
            body = err.read() if hasattr(err, "read") else b""
            if body:
                self.send_response(err.code)
                self.send_header("Content-Type", "application/json; charset=utf-8")
                self.send_header("Content-Length", str(len(body)))
                self.send_header("Cache-Control", "no-store")
                self.end_headers()
                self.wfile.write(body)
            else:
                self._send_json(err.code, {"error": "upstream_http_error", "message": str(err)})
        except URLError as err:
            self._send_json(502, {"error": "proxy_request_failed", "message": str(err)})

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path == "/api/gag2-stock.json":
            self._proxy_stock()
            return
        super().do_GET()


if __name__ == "__main__":
    server = ThreadingHTTPServer((HOST, PORT), Handler)
    print(f"GAG2 dev server running at http://{HOST}:{PORT}")
    print("Serves static tracker files and proxies /api/gag2-stock.json")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        server.server_close()
